README.md
# OctoPrint-TuyaSmartplug

Work based on [OctoPrint-TPLinkSmartplug](https://github.com/jneilliii/OctoPrint-TPLinkSmartplug) and [python-tuya](https://github.com/clach04/python-tuya).
This plugin controlls [Tuya-based](https://en.tuya.com/) SmartPlugs.

##  Screenshots
![screenshot](screenshot.png)

![screenshot](settings.png)

![screenshot](plugeditor.png)

## Setup

Install via the bundled [Plugin Manager](https://github.com/foosel/OctoPrint/wiki/Plugin:-Plugin-Manager)
or manually using this URL:

    https://github.com/ziirish/OctoPrint-TuyaSmartplug/archive/master.zip


## Preparatory work

In order to be able to interact with your Tuya smart plugs, you need to retrieve
both the `Device ID` and the `Local Key`. You'll find information to get those
in the [python-tuya wiki](https://github.com/clach04/python-tuya/wiki).

## Configuration

Once installed go into settings and enter the ip address for your TP-Link Smartplug device. Adjust additional settings as needed.

## Settings Explained
- **IP**
  - IP or hostname of the plug to control.
- **Label**
  - Label to use for title attribute on hover over button in navbar. Please note this settings is mandatory and should be unique across your plugs.
- **Icon Class**
  - Class name from [fontawesome](http://fontawesome.io/3.2.1/cheatsheet/) to use for icon on button.
- **Device ID**
  - Plug ID.
- **Local Key**
  - Local key to cypher data.
- **Plug Slot**
  - In case you have multiple slot, you can specify which one to interact with.
- **Warning Prompt**
  - Always warn when checked.
- **Warn While Printing**
  - Will only warn when printer is printing.
- **Use Countdown Timers**
  - Uses the plug's built in countdown timer rule to postpone the power on/off by configured delay in seconds.
- **GCODE Trigger**
  - When checked this will enable the processing of M80 and M81 commands from gcode to power on/off plug.  Syntax for gcode command is M80/M81 followed by hostname/ip.  For example if your plug is 192.168.1.2 your gcode command would be **M80 192.168.1.2**
  - You can also use the custom gcode commands `@TUYAON` and `@TUYAOFF` followed by the IP address of the plug.  This option will only work for plugs with GCODE processing enabled.  For example if your plug is 192.168.1.2 your gcode command would be **@TUYAON 192.168.1.2**
- **Auto Connect**
  - Automatically connect to printer after plug is powered on.
  - Will wait for number of seconds configured in **Auto Connect Delay** setting prior to attempting connection to printer.
- **Auto Disconnect**
  - Automatically disconnect printer prior to powering off the plug.
  - Will wait for number of seconds configured in **Auto Disconnect Delay** prior to powering off the plug.
- **Run System Command After On**
  - When checked will run system command configured in **System Command On** setting after a delay in seconds configured in **System Command On Delay**.
- **Run System Command Before Off**
  - When checked will run system command configured in **System Command Off** setting after a delay in seconds configured in **System Command Off Delay**.

## Support jneilliii Efforts
Most of the code used in this plugin has been written by
[jneilliii](https://github.com/jneilliii) so if you want to support someone,
you can support his work.


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.0] - NOT RELEASED YET
### Added
- Python 3 compatibility (#10 and #19)

### Fixed
- Embed pytuya code to avoid installation issues (#11, #15 and #20)

## [0.2.0] - 2019-10-21
### Added
- Support protocol 3.3 for newer hardware (#5)

## [0.1.0] - 2018-10-21
### Added
- Initial release based on [OctoPrint-TPLinkSmartplug by jneilliii](https://github.com/jneilliii/OctoPrint-TPLinkSmartplug).

[0.1.0]: https://github.com/ziirish/OctoPrint-TuyaSmartplug/tree/0.1.0
[0.2.0]: https://github.com/ziirish/OctoPrint-TuyaSmartplug/tree/0.2.0
[0.3.0]: https://github.com/ziirish/OctoPrint-TuyaSmartplug/tree/0.3.0


octoprint_tuyasmartplug/static/js/tuyasmartplug.js
/*
 * View model for OctoPrint-TuyaSmartplug
 *
 * Author: ziirish
 * License: AGPLv3
 */
$(function() {
    function tuyasmartplugViewModel(parameters) {
        var self = this;

        self.settings = parameters[0];
		self.loginState = parameters[1];

		self.arrSmartplugs = ko.observableArray();
		self.isPrinting = ko.observable(false);
		self.selectedPlug = ko.observable();
		self.processing = ko.observableArray([]);

		self.onBeforeBinding = function() {
			self.arrSmartplugs(self.settings.settings.plugins.tuyasmartplug.arrSmartplugs());
        }

		self.onAfterBinding = function() {
			self.checkStatuses();
		}

        self.onEventSettingsUpdated = function(payload) {
			self.arrSmartplugs(self.settings.settings.plugins.tuyasmartplug.arrSmartplugs());
		}

		self.onEventPrinterStateChanged = function(payload) {
			if (payload.state_id == "PRINTING" || payload.state_id == "PAUSED"){
				self.isPrinting(true);
			} else {
				self.isPrinting(false);
			}
		}

		self.cancelClick = function(data) {
			self.processing.remove(data.label());
		}

		self.editPlug = function(data) {
			self.selectedPlug(data);
			$("#TuyaPlugEditor").modal("show");
		}

		self.addPlug = function() {
			self.selectedPlug({'ip':ko.observable(''),
									'id':ko.observable(''),
									'slot':ko.observable(1),
									'localKey':ko.observable(''),
									'label':ko.observable(''),
									'icon':ko.observable('icon-bolt'),
									'displayWarning':ko.observable(true),
									'v33':ko.observable(false),
									'warnPrinting':ko.observable(false),
									'gcodeEnabled':ko.observable(false),
									'gcodeOnDelay':ko.observable(0),
									'gcodeOffDelay':ko.observable(0),
									'autoConnect':ko.observable(true),
									'autoConnectDelay':ko.observable(10.0),
									'autoDisconnect':ko.observable(true),
									'autoDisconnectDelay':ko.observable(0),
									'sysCmdOn':ko.observable(false),
									'sysRunCmdOn':ko.observable(''),
									'sysCmdOnDelay':ko.observable(0),
									'sysCmdOff':ko.observable(false),
									'sysRunCmdOff':ko.observable(''),
									'sysCmdOffDelay':ko.observable(0),
									'currentState':ko.observable('unknown'),
									'btnColor':ko.observable('#808080'),
									'useCountdownRules':ko.observable(false),
									'countdownOnDelay':ko.observable(0),
									'countdownOffDelay':ko.observable(0)});
			self.settings.settings.plugins.tuyasmartplug.arrSmartplugs.push(self.selectedPlug());
			$("#TuyaPlugEditor").modal("show");
		}

		self.removePlug = function(row) {
			self.settings.settings.plugins.tuyasmartplug.arrSmartplugs.remove(row);
		}

		self.onDataUpdaterPluginMessage = function(plugin, data) {
            if (plugin != "tuyasmartplug") {
                return;
            }

			plug = ko.utils.arrayFirst(self.settings.settings.plugins.tuyasmartplug.arrSmartplugs(),function(item){
				return item.label() === data.label;
				}) || {'label':data.label,'currentState':'unknown','btnColor':'#808080'};

			if (plug.currentState != data.currentState) {
				plug.currentState(data.currentState)
				switch(data.currentState) {
					case "on":
						break;
					case "off":
						break;
					default:
						new PNotify({
							title: 'Tuya Smartplug Error',
							text: 'Status ' + plug.currentState() + ' for ' + plug.ip() + '. Double check IP Address\\Hostname in tuyasmartplug Settings.',
							type: 'error',
							hide: true
							});
				self.settings.saveData();
				}
			}
			self.processing.remove(data.label);
        };

		self.toggleRelay = function(data) {
			self.processing.push(data.label());
			switch(data.currentState()){
				case "on":
					self.turnOff(data);
					break;
				case "off":
					self.turnOn(data);
					break;
				default:
					self.checkStatus(data.label());
			}
		}

		self.turnOn = function(data) {
/* 			if(data.sysCmdOn()){
				setTimeout(function(){self.sysCommand(data.sysRunCmdOn())},data.sysCmdOnDelay()*1000);
			} */
			self.sendTurnOn(data);
		}

		self.sendTurnOn = function(data) {
            $.ajax({
                url: API_BASEURL + "plugin/tuyasmartplug",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "turnOn",
					label: data.label()
                }),
                contentType: "application/json; charset=UTF-8"
            });
        };

    	self.turnOff = function(data) {
			if((data.displayWarning() || (self.isPrinting() && data.warnPrinting())) && !$("#tuyasmartplugWarning").is(':visible')){
				self.selectedPlug(data);
				$("#tuyasmartplugWarning").modal("show");
			} else {
				$("#tuyasmartplugWarning").modal("hide");
/* 				if(data.sysCmdOff()){
					setTimeout(function(){self.sysCommand(data.sysRunCmdOff())},data.sysCmdOffDelay()*1000);
				} */
				self.sendTurnOff(data);
			}
        };

		self.sendTurnOff = function(data) {
			$.ajax({
			url: API_BASEURL + "plugin/tuyasmartplug",
			type: "POST",
			dataType: "json",
			data: JSON.stringify({
				command: "turnOff",
				label: data.label()
			}),
			contentType: "application/json; charset=UTF-8"
			});
		}

		self.checkStatus = function(plugLabel) {
            $.ajax({
                url: API_BASEURL + "plugin/tuyasmartplug",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "checkStatus",
					label: plugLabel
                }),
                contentType: "application/json; charset=UTF-8"
            }).done(function(){
				self.settings.saveData();
				});
        };

		self.disconnectPrinter = function() {
            $.ajax({
                url: API_BASEURL + "plugin/tuyasmartplug",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "disconnectPrinter"
                }),
                contentType: "application/json; charset=UTF-8"
            });
		}

		self.connectPrinter = function() {
            $.ajax({
                url: API_BASEURL + "plugin/tuyasmartplug",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "connectPrinter"
                }),
                contentType: "application/json; charset=UTF-8"
            });
		}

/* 		self.sysCommand = function(sysCmd) {
            $.ajax({
                url: API_BASEURL + "plugin/tuyasmartplug",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "sysCommand",
					cmd: sysCmd
                }),
                contentType: "application/json; charset=UTF-8"
            });
		} */

		self.checkStatuses = function() {
			ko.utils.arrayForEach(self.settings.settings.plugins.tuyasmartplug.arrSmartplugs(),function(item){
				if(item.label() !== "") {
					console.log("checking " + item.label())
					self.checkStatus(item.label());
				}
			});
			if (self.settings.settings.plugins.tuyasmartplug.pollingEnabled()) {
				setTimeout(function() {self.checkStatuses();}, (parseInt(self.settings.settings.plugins.tuyasmartplug.pollingInterval(),10) * 60000));
			};
        };
    }

    // view model class, parameters for constructor, container to bind to
    OCTOPRINT_VIEWMODELS.push([
        tuyasmartplugViewModel,

        // e.g. loginStateViewModel, settingsViewModel, ...
        ["settingsViewModel","loginStateViewModel"],

        // "#navbar_plugin_tuyasmartplug","#settings_plugin_tuyasmartplug"
        ["#navbar_plugin_tuyasmartplug","#settings_plugin_tuyasmartplug"]
    ]);
});


octoprint_tuyasmartplug/__init__.py
# coding=utf-8
from __future__ import absolute_import

import octoprint.plugin
from octoprint.server import user_permission
import socket
import json
import logging
import os
import re
import threading
import time

from octoprint_tuyasmartplug.utils import pytuya


class tuyasmartplugPlugin(
    octoprint.plugin.SettingsPlugin,
    octoprint.plugin.AssetPlugin,
    octoprint.plugin.TemplatePlugin,
    octoprint.plugin.SimpleApiPlugin,
    octoprint.plugin.StartupPlugin,
):
    def __init__(self):
        self._logger = logging.getLogger("octoprint.plugins.tuyasmartplug")
        self._tuyasmartplug_logger = logging.getLogger(
            "octoprint.plugins.tuyasmartplug.debug"
        )

    # ~~ StartupPlugin mixin

    def on_startup(self, host, port):
        # setup customized logger
        from octoprint.logging.handlers import CleaningTimedRotatingFileHandler

        tuyasmartplug_logging_handler = CleaningTimedRotatingFileHandler(
            self._settings.get_plugin_logfile_path(postfix="debug"),
            when="D",
            backupCount=3,
        )
        tuyasmartplug_logging_handler.setFormatter(
            logging.Formatter("[%(asctime)s] %(levelname)s: %(message)s")
        )
        tuyasmartplug_logging_handler.setLevel(logging.DEBUG)

        self._tuyasmartplug_logger.addHandler(tuyasmartplug_logging_handler)
        self._tuyasmartplug_logger.setLevel(
            logging.DEBUG
            if self._settings.get_boolean(["debug_logging"])
            else logging.INFO
        )
        self._tuyasmartplug_logger.propagate = False

    def on_after_startup(self):
        self._logger.info("TuyaSmartplug loaded!")

    # ~~ SettingsPlugin mixin

    def get_settings_defaults(self):
        return dict(
            debug_logging=False,
            arrSmartplugs=[
                {
                    "ip": "",
                    "id": "",
                    "slot": 1,
                    "localKey": "",
                    "label": "",
                    "icon": "icon-bolt",
                    "displayWarning": True,
                    "warnPrinting": False,
                    "gcodeEnabled": False,
                    "v33": False,
                    "gcodeOnDelay": 0,
                    "gcodeOffDelay": 0,
                    "autoConnect": True,
                    "autoConnectDelay": 10.0,
                    "autoDisconnect": True,
                    "autoDisconnectDelay": 0,
                    "sysCmdOn": False,
                    "sysRunCmdOn": "",
                    "sysCmdOnDelay": 0,
                    "sysCmdOff": False,
                    "sysRunCmdOff": "",
                    "sysCmdOffDelay": 0,
                    "currentState": "unknown",
                    "btnColor": "#808080",
                    "useCountdownRules": False,
                    "countdownOnDelay": 0,
                    "countdownOffDelay": 0,
                }
            ],
            pollingInterval=15,
            pollingEnabled=False,
        )

    def get_settings_restricted_paths(self):
        return dict(
            admin=[
                [
                    "arrSmartplugs",
                ],
            ]
        )

    def on_settings_save(self, data):
        old_debug_logging = self._settings.get_boolean(["debug_logging"])

        octoprint.plugin.SettingsPlugin.on_settings_save(self, data)

        new_debug_logging = self._settings.get_boolean(["debug_logging"])
        if old_debug_logging != new_debug_logging:
            if new_debug_logging:
                self._tuyasmartplug_logger.setLevel(logging.DEBUG)
            else:
                self._tuyasmartplug_logger.setLevel(logging.INFO)

    def get_settings_version(self):
        return 3

    def on_settings_migrate(self, target, current=None):
        if current is None or current < self.get_settings_version():
            # Reset plug settings to defaults.
            self._logger.debug("Resetting arrSmartplugs for tuyasmartplug settings.")
            self._settings.set(
                ["arrSmartplugs"], self.get_settings_defaults()["arrSmartplugs"]
            )

    # ~~ AssetPlugin mixin

    def get_assets(self):
        return dict(js=["js/tuyasmartplug.js"], css=["css/tuyasmartplug.css"])

    # ~~ TemplatePlugin mixin

    def get_template_configs(self):
        return [
            dict(type="navbar", custom_bindings=True),
            dict(type="settings", custom_bindings=True),
        ]

    # ~~ SimpleApiPlugin mixin

    def turn_on(self, pluglabel):
        self._tuyasmartplug_logger.debug("Turning on %s." % pluglabel)
        if self.is_turned_on(pluglabel=pluglabel):
            self._tuyasmartplug_logger.debug("Plug %s already turned on" % pluglabel)
            self._plugin_manager.send_plugin_message(
                self._identifier, dict(currentState="on", label=pluglabel)
            )
            return
        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        self._tuyasmartplug_logger.debug(plug)
        if plug["useCountdownRules"]:
            chk = self.sendCommand(
                "countdown", plug["label"], int(plug["countdownOnDelay"])
            )
        else:
            chk = self.sendCommand("on", plug["label"])

        if chk is not False:
            self.check_status(plug["label"], chk)
            if plug["autoConnect"]:
                c = threading.Timer(
                    int(plug["autoConnectDelay"]), self._printer.connect
                )
                c.start()
            if plug["sysCmdOn"]:
                t = threading.Timer(
                    int(plug["sysCmdOnDelay"]), os.system, args=[plug["sysRunCmdOn"]]
                )
                t.start()
        else:
            self._plugin_manager.send_plugin_message(
                self._identifier, dict(currentState="unknown", label=pluglabel)
            )

    def turn_off(self, pluglabel):
        self._tuyasmartplug_logger.debug("Turning off %s." % pluglabel)
        if not self.is_turned_on(pluglabel=pluglabel):
            self._tuyasmartplug_logger.debug("Plug %s already turned off" % pluglabel)
            self._plugin_manager.send_plugin_message(
                self._identifier, dict(currentState="off", label=pluglabel)
            )
            return
        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        self._tuyasmartplug_logger.debug(plug)
        if plug["useCountdownRules"]:
            chk = self.sendCommand(
                "countdown", plug["label"], int(plug["countdownOffDelay"])
            )

        if plug["sysCmdOff"]:
            t = threading.Timer(
                int(plug["sysCmdOffDelay"]), os.system, args=[plug["sysRunCmdOff"]]
            )
            t.start()
        if plug["autoDisconnect"]:
            self._printer.disconnect()
            time.sleep(int(plug["autoDisconnectDelay"]))

        if not plug["useCountdownRules"]:
            chk = self.sendCommand("off", plug["label"])

        if chk is not False:
            self.check_status(plug["label"], chk)
        else:
            self._plugin_manager.send_plugin_message(
                self._identifier, dict(currentState="unknown", label=pluglabel)
            )

    def check_status(self, pluglabel, resp=None):
        self._tuyasmartplug_logger.debug("Checking status of %s." % pluglabel)
        if pluglabel != "":
            response = resp or self.sendCommand("info", pluglabel)
            if response is False:
                self._plugin_manager.send_plugin_message(
                    self._identifier, dict(currentState="unknown", label=pluglabel)
                )
            else:
                self._plugin_manager.send_plugin_message(
                    self._identifier,
                    dict(
                        currentState=(
                            "on" if self.is_turned_on(response, pluglabel) else "off"
                        ),
                        label=pluglabel,
                    ),
                )

    def is_turned_on(self, data=None, pluglabel=None):
        if data is None and pluglabel:
            data = self.sendCommand("info", pluglabel)

        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        return data and plug and data.get("dps", {}).get(str(plug["slot"]))

    def get_api_commands(self):
        return dict(turnOn=["label"], turnOff=["label"], checkStatus=["label"])

    def on_api_command(self, command, data):
        if not user_permission.can():
            from flask import make_response

            return make_response("Insufficient rights", 403)

        if command == "turnOn":
            self.turn_on("{label}".format(**data))
        elif command == "turnOff":
            self.turn_off("{label}".format(**data))
        elif command == "checkStatus":
            self.check_status("{label}".format(**data))

    # ~~ Utilities

    def plug_search(self, lst, key, value):
        for item in lst:
            if item[key] == value:
                return item

    def sendCommand(self, cmd, pluglabel, args=None, tries=1):
        self._tuyasmartplug_logger.debug("Sending command: %s to %s" % (cmd, pluglabel))
        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        device = pytuya.OutletDevice(plug["id"], plug["ip"], plug["localKey"])
        if plug.get("v33"):
            device.version = 3.3

        commands = {
            "info": ("status", None),
            "on": ("set_status", True),
            "off": ("set_status", False),
            "countdown": ("set_timer", None),
        }

        try:
            command, arg = commands[cmd]
            func = getattr(device, command, None)
            if not func:
                self._tuyasmartplug_logger.debug("No such command '%s'" % command)
                return False
            if args:
                func(args)
            elif arg is not None:
                func(arg, plug["slot"])
            else:
                func()
            time.sleep(0.5)
            ret = device.status()
            self._tuyasmartplug_logger.debug("Status: %s" % str(ret))
            return ret
        except socket.error as e:
            if e.errno == 104:
                if tries <= 3:
                    self._tuyasmartplug_logger.debug(
                        "Connection refused... Trying again soon"
                    )
                    time.sleep(1)
                    return self.sendCommand(cmd, pluglabel, args, tries + 1)
                self._tuyasmartplug_logger.debug("Too many failed attempts")
                return False
            self._tuyasmartplug_logger.debug("Network error")
            return False
        except:
            self._tuyasmartplug_logger.debug(
                "Something went wrong while running the command"
            )
            return False

    # ~~ Gcode processing hook

    def gcode_turn_off(self, plug):
        if plug["warnPrinting"] and self._printer.is_printing():
            self._logger.info(
                "Not powering off %s because printer is printing." % plug["label"]
            )
        else:
            self.turn_off(plug["label"])

    def processGCODE(self, comm_instance, phase, cmd, cmd_type, gcode, *args, **kwargs):
        if gcode:
            if cmd.startswith("M80"):
                name = re.sub(r"^M80\s?", "", cmd)
                self._tuyasmartplug_logger.debug(
                    "Received M80 command, attempting power on of %s." % name
                )
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]), "ip", name
                )
                if not plug:
                    plug = self.plug_search(
                        self._settings.get(["arrSmartplugs"]), "label", name
                    )
                self._tuyasmartplug_logger.debug(plug)
                if plug["gcodeEnabled"]:
                    t = threading.Timer(
                        int(plug["gcodeOnDelay"]), self.turn_on, args=[plug["label"]]
                    )
                    t.start()
                return
            elif cmd.startswith("M81"):
                name = re.sub(r"^M81\s?", "", cmd)
                self._tuyasmartplug_logger.debug(
                    "Received M81 command, attempting power off of %s." % name
                )
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]), "ip", name
                )
                if not plug:
                    plug = self.plug_search(
                        self._settings.get(["arrSmartplugs"]), "label", name
                    )
                self._tuyasmartplug_logger.debug(plug)
                if plug["gcodeEnabled"]:
                    t = threading.Timer(
                        int(plug["gcodeOffDelay"]), self.gcode_turn_off, args=[plug]
                    )
                    t.start()
                return
            else:
                return

        elif cmd.startswith("@TUYAON"):
            name = re.sub(r"^@TUYAON\s?", "", cmd)
            self._tuyasmartplug_logger.debug(
                "Received @TUYAON command, attempting power on of %s." % name
            )
            plug = self.plug_search(self._settings.get(["arrSmartplugs"]), "ip", name)
            if not plug:
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]), "label", name
                )
            self._tuyasmartplug_logger.debug(plug)
            if plug["gcodeEnabled"]:
                t = threading.Timer(
                    int(plug["gcodeOnDelay"]), self.turn_on, args=[plug["label"]]
                )
                t.start()
            return None
        elif cmd.startswith("@TUYAOFF"):
            name = re.sub(r"^@TUYAOFF\s?", "", cmd)
            self._tuyasmartplug_logger.debug(
                "Received TUYAOFF command, attempting power off of %s." % name
            )
            plug = self.plug_search(self._settings.get(["arrSmartplugs"]), "ip", name)
            if not plug:
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]), "label", name
                )
            self._tuyasmartplug_logger.debug(plug)
            if plug["gcodeEnabled"]:
                t = threading.Timer(
                    int(plug["gcodeOffDelay"]), self.gcode_turn_off, args=[plug]
                )
                t.start()
            return None

    # ~~ Softwareupdate hook

    def get_update_information(self):
        # Define the configuration for your plugin to use with the Software Update
        # Plugin here. See https://github.com/foosel/OctoPrint/wiki/Plugin:-Software-Update
        # for details.
        return dict(
            tuyasmartplug=dict(
                displayName="Tuya Smartplug",
                displayVersion=self._plugin_version,
                # version check: github repository
                type="github_release",
                user="ziirish",
                repo="OctoPrint-TuyaSmartplug",
                current=self._plugin_version,
                # update method: pip
                pip="https://github.com/ziirish/OctoPrint-TuyaSmartplug/archive/{target_version}.zip",
            )
        )


# If you want your plugin to be registered within OctoPrint under a different name than what you defined in setup.py
# ("OctoPrint-PluginSkeleton"), you may define that here. Same goes for the other metadata derived from setup.py that
# can be overwritten via __plugin_xyz__ control properties. See the documentation for that.
__plugin_name__ = "Tuya Smartplug"
__plugin_version__ = "0.3.0"
__plugin_pythoncompat__ = ">=2.7,<4"


def __plugin_load__():
    global __plugin_implementation__
    __plugin_implementation__ = tuyasmartplugPlugin()

    global __plugin_hooks__
    __plugin_hooks__ = {
        "octoprint.comm.protocol.gcode.queuing": __plugin_implementation__.processGCODE,
        "octoprint.plugin.softwareupdate.check_config": __plugin_implementation__.get_update_information,
    }
